CSS AtRules, Pseudo elements and Pseudo selector code hints

swmitra opened the pull request:
This PR adds CSS restricted block and pseudo selector/element code hints. In the implementation, priority of the hint providers have been kept at 0 while the priority of the default CSS code hint provider has been increased to 1. This has been done to ensure no additional overhead in case of CSS property:value code hints. Attached some screen shots of the corresponding code hints for illustration.
* @rules 
![atrules](https://cloud.githubusercontent.com/assets/12087205/24852890/bffe0b48-1df6-11e7-8c89-7401132113ca.png)

* Pseudo selectors 
![pseudoselector](https://cloud.githubusercontent.com/assets/12087205/24852854/a5fb3888-1df6-11e7-9884-b967169a6692.png)

* Pseudo elements 
![pseudoelements](https://cloud.githubusercontent.com/assets/12087205/24852873/b239148a-1df6-11e7-98be-5a78abb5319e.png)

Copying @ingorichter @petetnt @ficristo @zaggino @peterflynn  for code review.

Note: Additional description have been captured in the json def files to enable description tooltip at a later point of time.

swmitra committed:
AtRules, Pseudo elements and Pseudo selector code hints

petetnt commented on src/extensions/default/CSSCodeHints/AtRuleCodeHints.js:
```diff
@@ -0,0 +1,119 @@
+        if (this.filter) {
+
+            // Filter the property list based on the token string
+            var result = $.map(Object.keys(AtRules), function (key, value) {
```
Would prefer `Array.prototype.map` here, maybe with something like:

``` js
Object.keys(AtRules).map(function (key) {
  if (key.indexOf(token.string) === 0) {
    return key
  }
}).filter(function (key) {
  return key;
}).sort();
```

petetnt commented on src/extensions/default/CSSCodeHints/PseudoSelectorHints.js:
```diff
@@ -0,0 +1,171 @@
+        this.editor = editor;
+
+        // Check if we are at ':' pseudo rule or in 'variable-3' 'def' context
+        if (token.state.state === "pseudo" || token.type === "variable-3") {
```
`return token.state.state === "pseudo" || token.type === "variable-3"`

petetnt commented on src/extensions/default/CSSCodeHints/PseudoSelectorHints.js:
```diff
@@ -0,0 +1,171 @@
+
+        if (this.context !== -1) {
+            // Filter the property list based on the token string
+            var result = $.map(this.context === TOKEN_TYPE_PSEUDO_SELECTOR ? Object.keys(PseudoRules.selectors) : Object.keys(PseudoRules.elements), function (value, key) {
```
Prefer `Array.prototype.map`, maybe do the context check first.

``` js
var keys = this.context === TOKEN_TYPE_PSEUDO_SELECTOR ? Object.keys(PseudoRules.selectors) : Object.keys(PseudoRules.elements);

var result = keys.map(...)
```

petetnt commented on src/extensions/default/CSSCodeHints/PseudoSelectorHints.js:
```diff
@@ -0,0 +1,171 @@
+                handleWideResults: false
+            };
+        } else {
+            return false;
```
The other similar function returns `null` but I guess it doesn't matter

petetnt commented on src/extensions/default/CSSCodeHints/PseudoSelectorHints.js:
```diff
@@ -0,0 +1,171 @@
+            this.editor.setCursorPos({line: cursor.line, ch: cursor.ch - 1});
+        }
+
+        return false;
```
Does this need to return anything?

petetnt commented on src/extensions/default/CSSCodeHints/PseudoSelectorHints.js:
```diff
@@ -0,0 +1,171 @@
+        return false;
+    };
+
+    function _init() {
```
Empty method

ficristo replied:
AFAICT unused, remove.

petetnt commented on src/extensions/default/CSSCodeHints/PseudoSelectors.json:
```diff
@@ -0,0 +1,44 @@
+{
+    "selectors" :
+    {
+        "active" : { "desc": "Selects the active link"},
```
Normalize whitespace in this file between the keys and values

petetnt commented on src/extensions/default/CSSCodeHints/main.js:
```diff
@@ -468,7 +472,7 @@ define(function (require, exports, module) {
     AppInit.appReady(function () {
         var cssPropHints = new CssPropHints();
-        CodeHintManager.registerHintProvider(cssPropHints, ["css", "scss", "less"], 0);
+        CodeHintManager.registerHintProvider(cssPropHints, ["css", "scss", "less"], 1);
```
Q: What's the second parameter?

ficristo replied:
And why we need this change?

swmitra replied:
I have mentioned it in the opening note of the PR. 

> In the implementation, priority of the hint providers have been kept at 0 while the priority of the default CSS code hint provider has been increased to 1. This has been done to ensure no additional overhead in case of CSS property:value code hints.

To explain, the second parameter is for priority. CodeHintManager keeps the providers for a particular mode in order of priority, If there are priority collisions, then the lists are sorted based on loading order(this is not in our control). CodeHintManager first checks the current doc mode and then iterates over this sorted providers by calling 'hasHints', once it gets affirmative answer from a particular provider, it breaks there and selects that provider as the current session hint provider. By increasing the priority of the existing provider I am trying to ensure that, it gets the opportunity first to provide hints ( as hasHints() of this provider will be called first). This will ensure that, there is zero overhead on the current property:value hint provider.

petetnt replied:
@swmitra Sorry I already forgot about the opening note of PR 😸 thanks for the detailed explanation!

ficristo replied:
Same here...

swmitra replied:
👍

petetnt reviewed (changes requested):
Great job, some initial comments! The feature itself seems to work just fine!

ficristo commented:
This fixes #12486

ficristo commented on src/extensions/default/CSSCodeHints/AtRuleCodeHints.js:
```diff
@@ -0,0 +1,119 @@
+        }
+
+        // Check if we are at '@' rule 'def' context
+        if ((token.type === "def" && cmState.context.type === "at") || (token.type === "variable-2" && (cmState.context.type === "top" || cmState.context.type === "block"))) {
```
Split in two lines

ficristo commented on src/extensions/default/CSSCodeHints/AtRuleCodeHints.js:
```diff
@@ -0,0 +1,119 @@
+        this.filter = token.string;
+        this.token = token;
+
+        if (this.filter) {
```
I prefer an early return here:
```js
if (!this.filter) {
  return null;
}

var result...
```

ficristo commented on src/extensions/default/CSSCodeHints/AtRuleCodeHints.js:
```diff
@@ -0,0 +1,119 @@
+     *
+     * @return {boolean}
+     * Indicates whether the manager should follow hint insertion with an
+     * additional explicit hint request.
```
We should use an uniform style to doc comments: the description should be always inline or always at new line.
I think the current style is on the same line.

ficristo commented on src/extensions/default/CSSCodeHints/AtRuleCodeHints.js:
```diff
@@ -0,0 +1,119 @@
+    /**
+     * Inserts a given @<rule> hint into the current editor context.
+     *
+     * @param {string} hint
```
Wrong param name

ficristo commented on src/extensions/default/CSSCodeHints/AtRulesDef.json:
```diff
@@ -0,0 +1,10 @@
+    "@supports" : "A conditional group rule which will apply its content if the browser meets the criteria of the given condition.",
+    "@page" : "Describes the aspect of layout changes which will be applied when printing the document.",
+    "@font-face" : "Describes the aspect of an external font to be downloaded.",
+    "@keyframes"  : "Describes the aspect of intermediate steps in a CSS animation sequence."
```
In general I prefer no space before `:` and one after, but I think, in general, we use vertical align?

ficristo replied:
Curiosity: is it possible to translate these descriptions? I think it would be nice.

ficristo commented on src/extensions/default/CSSCodeHints/PseudoSelectorHints.js:
```diff
@@ -0,0 +1,171 @@
+
+
+/*jslint vars: true, plusplus: true, devel: true, nomen: true, indent: 4, maxerr: 50 */
+/*global define, brackets, $ */
```
As before, remove.

ficristo commented on src/extensions/default/CSSCodeHints/PseudoSelectorHints.js:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    // As we are only going to provide :<pseudo> name hints
+    // we should claim that we don't have hints for anything else
```
Use JSDoc and add the params

ficristo commented on src/extensions/default/CSSCodeHints/PseudoSelectorHints.js:
```diff
@@ -0,0 +1,171 @@
+        this.token = token;
+
+
+        if (this.context !== -1) {
```
Early return.

ficristo commented:
I briefly tested, but I cannot use the pseudo selectors in a SCSS file, but I can in a CSS one.

Do you have thought of using multiple extensions like suggested by @redmunds in https://github.com/adobe/brackets/issues/12486#issuecomment-257583129?
I don't like much the `require`s with side effects in main.js.
Do you think is possible to have AtRuleCodeHints, PseudoSelectorHints and CssPropHints files which exports their class and a main.js file with a single `AppInit.appReady` function that create an instance for every of the three classes?

swmitra committed:
Adress review comments and restructure using different extensions

swmitra commented:
@petent @ficristo I have addressed all the review comments in my last commit. Also extracted these providers as separate default extensions. Need your review once more 😄

ficristo commented:
In general LGTM, but I still cannot see pseudo selector hints for SCSS code.
Also add some tests.

I've use this snippet to test:
```scss
@charset "utf8";
/*comment*/

.test1: {
}

.test2:: {
}
```

swmitra commented:
@ficristo You probably have 'brackets-saas' extension installed. That extension registers 'SaasHints' provider with high priority and claims to provide hints ( 'hasHints' call returns true) even when it can't. 
Please disable that extension and try. It should work.

ficristo commented:
I disabled all my extensions and I can still reproduce.
If you write `.test:` in a SCSS file the hints work.
If you write `.test:: {}`, move the cursor after the last `:` and use Ctrl-space to ask hints, nothing happen. But the same works in a CSS file.

swmitra commented:
@ficristo WIth the SCSS tokenizer in CM2, the token identifier is not able to detect the pseudo context in the second scenario you have mentioned. I have filed an issue with CM https://github.com/codemirror/CodeMirror/issues/4688 just now. Lets wait for a response from @marijnh.

swmitra commented:
@ficristo Please refer to the CM issue and the discussion at https://github.com/codemirror/CodeMirror/issues/4688, we might have to live with this issue for sometime 😞 . But fortunately the moment you put the first char after `:`, you will get hints if the typed characters matches any of them. I tried to get some workaround with regex, but it's not going to work anyway as CM matches `/:\s*\{/` and combines them to a single token. So distinguishing between `/:\s*\{/` and `/::\s*\{/` is virtually impossible for me (unless someone can figure out a hack not dependent on the CM token).

marijnh commented:
Just figuring out what the SCSS mode is doing and fixing it is going to be a lot easier than trying to work around this from the outside.

swmitra commented:
Thanks @marijnh for joining in 👍 
I was just going through some of the syntax's in SCSS and most likely need to spend some more time to abstractly make sure that I don't break anything while fixing the CM CSS mode parser 😄

ficristo commented:
According to http://sass-lang.com/documentation/file.SASS_REFERENCE.html#nested_properties you can "split" properties like this:
```scss
.test {
  font: {
    family: serif;
  }
}
```
instead of writing
```scss
.test {
  font-family: serif;
}
```
I don't know what we can do about it for our case.

@swmitra is the "problem" also present with `::` ?

swmitra commented:
yes @ficristo. The regex used in the mode parser matches both the cases. This is what it does for punctuation `:`  handling in mode/css/css.js @line 752 - 
```javascript
":": function(stream) {
        if (stream.match(/\s*\{/))
          return [null, "{"];
        return false;
      }
```
Now if I remove this or try to provide a punctuation handler without wrapping the `/\s{/` , our property:value handling will go for a toss 😞 in the following SCSS construct 

```scss
.test {
  font: {
    family: serif;
  }
}
```
@marijnh Badly need your input here...

swmitra commented:
@marijnh I see that the following https://github.com/codemirror/CodeMirror/commit/2c46d1b45af0baba87804f451c566de256c8d390 fixes the issue 👍

swmitra committed:
Handle SCSS mode seperately

swmitra commented:
@ficristo The issue in SCSS mode is fixed now using CM https://github.com/codemirror/CodeMirror/commit/2c46d1b45af0baba87804f451c566de256c8d390 and my last commit. Can you please check it once?

ficristo commented:
To see the last changes in action we need to wait for a CodeMirror upgrade, is it correct?
Also this PR needs some tests.

swmitra commented:
Yes @ficristo, I will upgrade CM once a release is available with the mentioned commit. I am writing tests for both the extensions...

ingorichter commented on src/extensions/default/CSSAtRuleCodeHints/main.js:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        // Filter the property list based on the token string
+        var result = Object.keys(AtRules).map(function (key) {
```
I might be wrong, but I think `filter` would be more appropriate than `map` for the first loop. This would also make the following `filter` loop unnecessary.

ingorichter commented on src/extensions/default/CSSPseudoSelectorHints/PseudoSelectors.json:
```diff
@@ -0,0 +1,44 @@
+        "active":              {"desc": "Selects the active link"},
+        "checked":             {"desc": "Selects every checked <input> element"},
+        "disabled":            {"desc": "Selects every disabled <input> element"},
+        "empty"	:              {"desc": "Selects every element that has no children/text (including text nodes)"},
```
nit: formatting is a bit off

ingorichter commented on src/extensions/default/CSSPseudoSelectorHints/PseudoSelectors.json:
```diff
@@ -0,0 +1,44 @@
+    },
+    "elements" :
+    {
+        "after":        {"desc": "Insert something after the content of each element identified by this selctor"},
```
nit: `selector`

ingorichter commented on src/extensions/default/CSSPseudoSelectorHints/PseudoSelectors.json:
```diff
@@ -0,0 +1,44 @@
+    "elements" :
+    {
+        "after":        {"desc": "Insert something after the content of each element identified by this selctor"},
+        "before":       {"desc": "Insert something before the content of each element identified by this selctor"},
```
nit: same `selector` spelling

ingorichter replied:
basically all instances of `selctor` in this object ;-)

ingorichter commented on src/extensions/default/CSSPseudoSelectorHints/main.js:
```diff
@@ -0,0 +1,176 @@
+                return key;
+            }
+        }).filter(function (key) {
+            return key;
```
replacing the previous `map` call with `filter` should make this obsolete

ingorichter commented on src/extensions/default/CSSPseudoSelectorHints/main.js:
```diff
@@ -0,0 +1,176 @@
+        this.editor = editor;
+
+        // Check if we are at ':' pseudo rule or in 'variable-3' 'def' context
+        return token.state.state === "pseudo" || token.type === "variable-3" || token.string === PUNCTUATION_CHAR;
```
nit: this could be a function determining the state. The same statement (negated) is used in `getHints` and could be reused there. Plus, it makes testing this function in isolation easier.

ingorichter reviewed (commented):
A couple of nits. Otherwise LGTM

swmitra committed:
Merge branch 'master' of https://github.com/adobe/brackets into swmitra/EnhancedCSSCodeHints

swmitra committed:
Adding unit tests for the newly added extensions

swmitra committed:
Fix for eslint errors

swmitra commented:
@petetnt @ingorichter @ficristo Can you guys please have a look at this PR now? 
CM dependency has been resolved with #13331 and also I have addressed all review comments and added unit tests for all style modes (css, less, scss).

ficristo commented:
There is still the nit of @ingorichter to fix. No time to look again.

swmitra committed:
Extract pseudo context validation as a function
